File: split/newhyphexceptions.pas
procedure newhyphexceptions;
label
  21, 10, 40, 45;
var
  n: 0..64;
  j: 0..64;
  h: hyphpointer;
  k: strnumber;
  p: halfword;
  q: halfword;
  s, t: strnumber;
  u, v: poolpointer;
begin
  scanleftbrace;
  if eqtb[5313].int <= 0 then
    curlang := 0
  else if eqtb[5313].int > 255 then
    curlang := 0
  else
    curlang := eqtb[5313].int;
  {935:}
n := 0;
  p := 0;
  while true do
  begin
    getxtoken;
    21:
      case curcmd of
        11, 12, 68:{937:}
if curchr = 45 then{938:}

          begin
            if n < 63 then
            begin
              q := getavail;
              mem[q].hh.rh := p;
              mem[q].hh.lh := n;
              p := q;
            end;
          end{:938}

          else if eqtb[4239 + curchr].hh.rh = 0 then
          begin
            begin
              if interaction = 3 then;
              printnl(262);
              print(944);
            end;
            begin
              helpptr := 2;
              helpline[1] := 945;
              helpline[0] := 946;
            end;
            error;
          end
          else if n < 63 then
          begin
            n := n + 1;
            hc[n] := eqtb[4239 + curchr].hh.rh;
          end{:937}
;
        16:
        begin
          scancharnum;
          curchr := curval;
          curcmd := 68;
          goto 21;
        end;
        10, 2:
        begin
          if n > 1 then{939:}

          begin
            n := n + 1;
            hc[n] := curlang;
            begin
              if poolptr + n > poolsize then
                overflow(257, poolsize - initpoolptr);
            end;
            h := 0;
            for j := 1 to n do
            begin
              h := (h + h + hc[j]) mod 307;
              begin
                strpool[poolptr] := hc[j];
                poolptr := poolptr + 1;
              end;
            end;
            s := makestring;
            {940:}
if hyphcount = 307 then
              overflow(947, 307);
            hyphcount := hyphcount + 1;
            while hyphword[h] <> 0 do
            begin
              {941:}
k := hyphword[h];
              if (strstart[k + 1] - strstart[k]) < (strstart[s + 1] - strstart[s]) then
                goto 40;
              if (strstart[k + 1] - strstart[k]) > (strstart[s + 1] - strstart[s]) then
                goto 45;
              u := strstart[k];
              v := strstart[s];
              repeat
                if strpool[u] < strpool[v] then
                  goto 40;
                if strpool[u] > strpool[v] then
                  goto 45;
                u := u + 1;
                v := v + 1;
              until u = strstart[k + 1];
            40:
              q := hyphlist[h];
              hyphlist[h] := p;
              p := q;
              t := hyphword[h];
              hyphword[h] := s;
              s := t;
            45:
              {:941}
;
              if h > 0 then
                h := h - 1
              else
                h := 307;
            end;
            hyphword[h] := s;
            hyphlist[h] := p{:940}
;
          end{:939}
;
          if curcmd = 2 then
            goto 10;
          n := 0;
          p := 0;
        end;
        others:{936:}

        begin
          begin
            if interaction = 3 then;
            printnl(262);
            print(680);
          end;
          printesc(940);
          print(941);
          begin
            helpptr := 2;
            helpline[1] := 942;
            helpline[0] := 943;
          end;
          error;
        end{:936}

      end;
  end{:935}
;
  10: ;
end;

File: cpp/src/newhyphexceptions.cpp
#include "newhyphexceptions.h"
#include "scanleftbrace.h"
#include "getxtoken.h"
#include "getavail.h"
#include "impression.h"
#include "scancharnum.h"
#include "makestring.h"
#include "erreur.h"
#include "texte.h"

void newhyphexceptions(void)
{
	char n, j;
	hyphpointer h;
	halfword p;
	poolpointer u, v;
	scanleftbrace();
	curlang = cur_fam();
	if (curlang < 0 || curlang > 255)
		curlang = 0;
	n = 0;
	p = 0;
	getxtoken();
	while (true)
	{
		switch (curcmd)
		{
			case letter:
			case other_char:
			case char_given:
				if (curchr == '-')
				{
					if (n < 63)
					{
						auto q = getavail();
						link(q) = p;
						info(q) = n;
						p = q;
					}
				}
				else 
					if (lc_code(curchr) == 0)
						error("Not a letter", "Letters in \\hyphenation words must have \\lccode>0.\nProceed; I'll ignore the character I just read.");
					else 
						if (n < 63)
						{
							n++;
							hc[n] = lc_code(curchr);
						}
				break;
			case char_num:
				scancharnum();
				curchr = curval;
				curcmd = char_given;
				continue;
			case spacer:
			case right_brace:
				if (n > 1)
				{
					n++;
					hc[n] = curlang;
					str_room(n);
					h = 0;
					for (j = 1; j <= n; j++)
					{
						h = (2*h+hc[j])%hyph_size;
						append_char(hc[j]);
					}
					auto s = makestring();
					if (hyphcount == hyph_size)
						overflow("exception dictionary", hyph_size);
					hyphcount++;
					while (hyphword[h] != "")
					{
						if (hyphword[h].size() < s.size())
						{
							std::swap(hyphlist[h], p);
							std::swap(hyphword[h], s);
						}
						else 
							if (hyphword[h].size() == s.size())
							{
								u = 0;
								v = 0;
								bool label45 = false;
								do
								{
									if (hyphword[h][u] < s[v])
									{
										std::swap(hyphlist[h], p);
										std::swap(hyphword[h], s);
										label45 = true;
									}
									if (hyphword[h][u] > s[v])
										label45 = true;
									if (label45)
										break;
									u++;
									v++;
								} while (u < hyphword[h].size());
								if (!label45)
								{
									std::swap(hyphlist[h], p);
									std::swap(hyphword[h], s);
								}
							}
						if (h > 0)
							h--;
						else
							h = hyph_size;
					}
					hyphword[h] = s;
					hyphlist[h] = p;
				}
				if (curcmd == right_brace)
					return;
				n = 0;
				p = 0;
				break;
			default:
				error("Improper "+esc("hyphenation")+" will be flushed", "Hyphenation exceptions must contain only letters\nand hyphens. But continue; I'll forgive and forget.");
		}
		getxtoken();
	}
}
